controllers/machinepool_controller_noderef.go

```diff
@@ -0,0 +1,167 @@
+	}
+
+	// Check that the MachinePool has a valid ProviderID.
+	if machinepool.Spec.ProviderIDs == nil || len(machinepool.Spec.ProviderIDs) == 0 {
+		klog.Warningf("MachinePool %q in namespace %q doesn't have a valid ProviderID yet", machinepool.Name, machinepool.Namespace)
+		return nil
+	}
+
+	clusterClient, err := remote.NewClusterClient(r.Client, cluster, r.scheme)
+	if err != nil {
+		return err
+	}
+
+	err = r.deleteRetiredNodes(ctx, clusterClient, machinepool.Status.NodeRefs, machinepool.Spec.ProviderIDs)
+	if err != nil {
+		return err
+	}
+
+	// Get the Node reference.
+	nodeRefs, available, ready, err := r.getNodeReferences(ctx, clusterClient, machinepool.Spec.ProviderIDs)
+	if err != nil {
+		if err == ErrNodeNotFound {
+			return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 10 * time.Second},
+				"cannot assign NodeRefs to MachinePool %q in namespace %q, no matching Nodes", machinepool.Name, machinepool.Namespace)
+		}
+		klog.Errorf("Failed to assign NodeRefs to MachinePool %q in namespace %q: %v", machinepool.Name, machinepool.Namespace, err)
+		r.recorder.Event(machinepool, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
+		return err
+	}
+
+	machinepool.Status.AvailableReplicas = int32(available)
+	machinepool.Status.ReadyReplicas = int32(ready)
+	machinepool.Status.NodeRefs = nodeRefs
+	klog.Infof("Set MachinePool's (%q in namespace %q) NodeRefs to %q", machinepool.Name, machinepool.Namespace, machinepool.Status.NodeRefs)
+	r.recorder.Event(machinepool, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", fmt.Sprintf("%+v", machinepool.Status.NodeRefs))
+
+	if machinepool.Status.ReadyReplicas == 0 || len(nodeRefs) != int(machinepool.Status.ReadyReplicas) {
+		return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 10 * time.Second},
+			"NodeRefs != ReadyReplicas [%q != %q] for MachinePool %q in namespace %q", len(nodeRefs), machinepool.Status.ReadyReplicas, machinepool.Name, machinepool.Namespace)
+	}
+	return nil
+}
+
+func (r *MachinePoolReconciler) deleteRetiredNodes(ctx context.Context, client client.Client, nodeRefs []apicorev1.ObjectReference, providerIDs []string) error {
+	nodeRefsMap := make(map[string]*apicorev1.Node)
+	for _, nodeRef := range nodeRefs {
+		node := &corev1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: nodeRef.Name,
+			},
+		}
+		err := client.Get(ctx, types.NamespacedName{Name: nodeRef.Name}, node)
+		if err != nil {
+			return err
+		}
+		nodeRefsMap[node.Spec.ProviderID] = node
+	}
+	for _, providerID := range providerIDs {
+		delete(nodeRefsMap, providerID)
+	}
+	for _, node := range nodeRefsMap {
+		err := client.Delete(ctx, node)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *MachinePoolReconciler) getNodeReferences(ctx context.Context, client client.Client, providerIDs []string) ([]apicorev1.ObjectReference, int, int, error) {
+	var ready, available int
+	nodeRefsMap := make(map[string]apicorev1.Node)
+	listOpt := metav1.ListOptions{}
```

Comment:
I'm guessing you were copying this from MachineReconciler.getNodeReference, which isn't doing this correctly 😄. This is roughly how it should look:

```go
nodeList := apicorev1.NodeList{}
for {
  if err := c.List(ctx, &nodeList, client.Continue(nodeList.Continue)); err != nil {
    // check error
  }

  // do stuff

  if nodeList.Continue == "" {
    break
  }
}
```

Reply:
(I filed https://github.com/kubernetes-sigs/cluster-api/issues/2166 for MachineReconciler)